sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -254,24 +255,104 @@ object DecorrelateInnerQuery extends PredicateHelper {
       // is the attribute from the outer query. (See comment above in buildDomainAttrMap.)
       // We need to remap the attribute names used in the inner query (left hand side) to account
       // for the different names in each union child. We should not remap the attribute names used
       // in the outer query.
+      //
+      // Note: the reason we can't just use the original joinCond from when the DomainJoin was
+      // constructed is that constructing the DomainJoins happens much earlier than rewriting the
+      // DomainJoins into actual joins, with many optimization steps in
+      // between, which could change the attributes involved (e.g. CollapseProject).
       case EqualNullSafe(left: Attribute, right: Expression) =>
         EqualNullSafe(map.getOrElse(left, left), right)
-      case EqualTo(left: Attribute, right: Expression) =>
-        EqualTo(map.getOrElse(left, left), right)
+    }
+  }
+
+  /**
+   * This is to handle INTERSECT/EXCEPT DISTINCT which are rewritten to left semi/anti join in
+   * ReplaceIntersectWithSemiJoin and ReplaceExceptWithAntiJoin.
+   *
+   * To rewrite the domain join on the right side, we need to remap the attributes in the domain
+   * join cond, using the mapping between left and right sides in the semi/anti join cond.
+   *
+   * After DecorrelateInnerQuery, the domain join conds reference the output names of the
+   * INTERSECT/EXCEPT, which come from the left side. When rewriting the DomainJoin in the
+   * right child, we need to remap the domain attribute names to account for the different
+   * names in the left vs right child, similar to pushDomainConditionsThroughSetOperation.
+   * But after the rewrite to semi/anti join is performed, we instead need to do the remapping
+   * based on the semi/anti join cond which contains equi-joins between the left and right
+   * outputs.
+   *
+   * The domain conds will be like leftInner <=> outer, and the semi/anti join cond will be like
+   * leftInner <=> rightInner. We add additional domain conds rightInner <=> outer which are used
+   * to rewrite the right-side DomainJoin.
+   *
+   * Example: Take a query like:
+   * select * from t0 join lateral (
+   *   select a from t1 where b < x
+   *   intersect distinct
+   *   select b from t2 where c < y)
+   *
+   * Step 1:
+   * We have outer table with attributes [x#1, y#2] and after introducing DomainJoins the subquery
+   * is a Intersect where the left side has output [a#3, x#4, y#5] with DomainJoin [x#4, y#5]
+   * and the right side has output [b#6, x#7, y#8] with DomainJoin [x#7, y#8].
+   * The domain join conditions are x#4 <=> x#1 and y#5 <=> y#2, because the output of a set op
+   * uses the attribute names from the left side.
+   *
+   * Step 2:
+   * ReplaceIntersectWithSemiJoin runs and transforms the Intersect to
+   * Join LeftSemi, (((a#3 <=> b#6) AND (x#4 <=> x#7)) AND (y#5 <=> y#8))
```

Comment:
This join between t1,t2 happens for both intersect distinct and intersect all? It seems so, so maybe update the comment? 

also mention that Intersect(isall=?) will take care of distinct/all?

Reply:
The join rewrite only happens for intersect/except distinct. For ALL it gets rewritten to a UNION and goes through the existing pushDomainConditionsThroughSetOperation logic. You can find a more detailed description in the doc, I also edited the comment to add this.